### ISSUE TITLE
Request to merge the maven branch from ssill2/apg-java back to your repo

### ISSUE THREAD

**Issue by ssill2** [ldthomas/apg-java#8]
Hi Lowell,
I did some very minor code changes to satisfy Fortify scans.  I was wondering if you wanted to make my maven branch a branch in your repo and I could just do pull requests if I make any more changes.  you could then choose if you wanted to merge them to master or not.

**Comment by ldthomas** [comment]
First of all let me say that I really like the way you have set up APG as a maven project and that I wouldn’t mind having it as a “maven” branch. If I were starting over, that’s the way I would set it up. But it’s been what it’s been for 10 years and I’m not actively developing it, so I don’t want to upset that apple cart. So even if I added your maven branch I wouldn’t have any plans on ever merging it with master unless I got back into active Java development.

As for the changes you’ve made to satisfy the Fortify scans, those I would merge into master if there were a way to do that. Unfortunately, it looks to me like that would have to be a manual process on the master branch. 

I’m not a big GitHub expert. I’m not sure how to do a pull request that would bring in your maven branch without changing master. And to be honest I’m pretty skittish about really screwing things up by trying.

**Comment by ssill2** [comment]
I understand.   Since I've not made any changes to my fork of your master I THINK you should be able to bring my branch in as a "maven" branch in your repo without distrupting master at all.  At least that's how I understand git to work.

If that part can be accomplished without upsetting the apple cart in master, would you be open do doing it?  I can do some research on what the process might be if you're open to it.
I think changes to individual files would just need to be watched for and hand ported between master and maven.  Which I don't think is a huge deal given how little change goes on.

**Comment by ldthomas** [comment]
Yes, I'm interested. I know how to create a new branch and push that branch to GitHub _as_ a new branch, but pulling a forked branch that is not master/main is where I'm confused. Let me know what you find out.

**Comment by ssill2** [comment]
Yep, looking into it now.  Will let you know what I find out.

**Comment by ssill2** [comment]
so I've spoken about this to people more knowledgeable about this than myself.  They suggested going ahead and creating a branch of master called maven and to have me attempt do a pull request against that.  Thus changes would only go to that branch and not master.  would you be up for doing that?

**Comment by ldthomas** [comment]
I'm guessing that will work but I don't like to experiment with the live repository. It has quite a few forks and some of them are staying current so I don't want to disrupt that. How about if we test the concept first? I'll create a new repository "multibranch" with a few simple text files in it. You fork it and create a maven branch with a seriously different file structure. I will independently push up my own simple maven branch. Then you do a pull request on the maven branch. If that works we can repeat it with apg-java.

But It is getting a little late in the PM (EDT) for me to start something like that. If you are game I'll set it up Monday morning and we can give it a spin.

**Comment by ssill2** [comment]
yeah that's a good idea.   So basically new repo with a master.   Create your files and dirs.  Then branch and call it maven.  I'll fork and and make a bunch of changes in the the maven branch and then do a pull request and we'll see what happens! lol  Thanks

**Comment by ldthomas** [comment]
I’ve done some testing with multiple branches on a private repository and I think this will all work ok. At least when creating a pull request from one branch to another the process asks which branch you want to pull to. Main or master is the default but if there are multiple branches you can select the one you want. But I can’t fork my own repository so I’m not exactly sure what it looks like when creating a pull request from a fork but I’m guessing it will be similar. In any case, I’d like to test it once before making a major blunder on the real deal.

I’ve created a new repository “blackbird”. It is actually a copy of apg-java. My suggestion is to fork it and in your fork create a new branch “maven” and just copy and commit your maven branch to it. Let me know when you are ready and I will push up a new maven branch. (It may seem like a fine point, but I’m purposely waiting until you have forked _before_ pushing up a new maven branch because that is the sequence we will be using on apg-java.) Then see if you can create a pull request from your forked maven branch to my blackbird maven branch. I’ll merge the pull request and check that 1) main/master isn’t modified and 2) the new maven branch matches yours.

If that all works we can repeat with apg-java and I'll delete blackbird.

**Comment by ssill2** [comment]
so I think you want to create a branch of master in blackbird called maven.  Then I create the fork and make modifications to my copy of your maven branch and then submit a pull request.  At least that's how I understand the workflow. 

If this works then I think I can rename my maven branch of apg-java to something else and you can create another maven branch of apg-java and I can do a pull request to pull changes from my renamed maven branch to your maven branch.

Does that make sense?

**Comment by ldthomas** [comment]
I'm trying to recreate on blackbird exactly the situation we are facing with apg-java. I have a single master branch, you have a fork with a maven branch. That's where we are right now. I want to now create a maven branch in addition to master and have you create a pull request from your forked maven branch to my maven branch. That's what we want to do.

So, blackbird right now has a single main branch. If you fork and create a maven branch then we have with blackbird exactly the same situation as we currently do with apg-java. Then we are in position for me to create my maven branch and for you to create a pull request into it.

Not trying to be too confusing. Just trying to recreate our actual situation with apg-java. What do you think?

**Comment by ssill2** [comment]
So after speaking with some folks that are more knowledgeable than I about github and git in general I came away with the impression that git in general supports the idea but the github UI doesn't seem to support taking a branch I make and pushing it up to your repo as a pull request.  As you saw the pull request drop downs only let you pick master as the destination, not a non-existent branch.
So what we should test in blackbird is:
You create a branch of master called maven.
I fork the blackbird project with its two branches.
I create a branch of your master branch in my fork called something other than master, e.g. maven_work
I make changes in maven_work
I attempt to merge maven_work into maven locally.  remember maven_work was created from master(like in apg_java).  If I can merge maven_work in to my copy of your maven branch, I can then create a pull request for you to pull the state of maven in my fork back to your repo.
Does that make sense?

**Comment by ldthomas** [comment]
The thing that I'm missing here is that with apg-java you have already forked it without a maven branch. The question is, if I create a maven branch can you create a pull request into it. If so, I don't see the need for a third maven_fork branch.

**Comment by ldthomas** [comment]
maven_work I mean

**Comment by ssill2** [comment]
maven_work would go away in my local repo once I merged it to maven in my local repo and successfully did the pull request to have you pull the state of my copy of the maven branch to your repo.

Make sense.  From then on maven would just live as it's own branch in your repo and if I make more changes we can just do pull requests to get them in to your branch.

**Comment by ssill2** [comment]
since maven in your branch would be ancestrally related to master, you could merge it in at a later time or never also  it could just live there in parallel to master too evolving differently.

**Comment by ldthomas** [comment]
OK, whatever makes sense to you on your end. As long as you can create a pull request from any of your forked branches into my maven branch we should be good. I've pushed up a new maven branch to blackbird ( it's identical to main). I'll wait to see if I get a pull request into it and we will see how that goes.

**Comment by ssill2** [comment]
alright I'll fork that here in a bit and give it a shot.  Fingers crossed!

**Comment by ssill2** [comment]
I see you've done the pull.  Did that work ok?  I've renamed my maven branch in my fork to maven_work as to be out of the way if you create a maven branch in your repo.

**Comment by ldthomas** [comment]
I'll be damned. It worked. I'll push up a maven branch to apg-java and we can repeat there.

**Comment by ssill2** [comment]
I think we are both learning stuff lol.  Glad it worked!

**Comment by ldthomas** [comment]
OK. I'm really holding my breath now. I've created a maven branch on apg-java. Go ahead with a pull request from your fork and after I have double, _double_ checked that it is going to the maven branch I'll do the merge and we should be home free.

**Comment by ssill2** [comment]
PR should be ready for your review

**Comment by ldthomas** [comment]
Whoa! High five, dude! We did it.

**Comment by ssill2** [comment]
Nice!   should be easier from now on.  I've made about all the fortify related changes I'm comfortable making.  So many of them are BS findings.   I'll let you know if end up making any more changes that would be generally useful to folks.  Thanks for working with me on this! :)

**Comment by ldthomas** [comment]
My pleasure. When you need to update the maven branch it should just be a "simple" pull request. I'll close this out for now. Have fun and good luck with your endeavors.
